File: 2385.cpp
#include <tuple>

class Solution {
    int res = -1;
    int Start;
public:
    std::tuple<int, int> dfs(TreeNode *node) {
        if (!node) {
            return make_pair(-1, -1);
        }
        int l1, l2, r1, r2;
        std::tie(l1, l2) = dfs(node->left);
        std::tie(r1, r2) = dfs(node->right);
        int d=-1;
        if (node->val == Start) {
            res = max(l1, r1) + 1;
            d = 0;
        } else if (l2 >= 0) {
            res = max(res, r1 + l2 + 2);
            d = l2 + 1;
        } else if (r2 >= 0) {
            res = max(res, l1+r2 + 2);
            d = r2 + 1;
        }
        return std::make_tuple(max(l1, r1) + 1, d);  
    }

    int amountOfTime(TreeNode* root, int start) {
        Start = start;
        dfs(root);
        return res;
    }
};

File: Бабат/2385.py
class TreeNode:
     def __init__(self, val=0, left=None, right=None):
         self.val = val
         self.left = left
         self.right = right
class Solution:
    def amountOfTime(self, root: Optional[TreeNode], start: int) -> int:
        graph = defaultdict(list)

        def buildGraph(node):
            if node:
                if node.left:
                    graph[node.val].append(node.left.val)
                    graph[node.left.val].append(node.val)
                    buildGraph(node.left)
                if node.right:
                    graph[node.val].append(node.right.val)
                    graph[node.right.val].append(node.val)
                    buildGraph(node.right)

        buildGraph(root)

        queue = deque([(start, 0)])
        visited = set([start])
        minutes = 0

        while queue:
            node, time = queue.popleft()
            minutes = max(minutes, time)

            for neighbor in graph[node]:
                if neighbor not in visited:
                    visited.add(neighbor)
                    queue.append((neighbor, time + 1))

        return minutes
